changes/21340-improve-nano-enrollments-last-seen-at-update

```diff
@@ -0,0 +1 @@
+* Improve performance of updating the `nano_enrollments.last_seen_at` timestamp of Apple MDM devices by an order of magnitude under load.
```

Comment:
Based on some quick searching in the repo, I don't really see where `last_seen_at` is being used by nano for any operations (so the column could in theory be removed altogether without impacting nano functionality). How does that align with your understanding?

Reply:
lol! Wow, can I get the last 3-4 days back :D 

It seems like you may be right, it looks like it's only used in the touchless migration, but then again it seems like it's only to set it? https://github.com/fleetdm/fleet/blob/6b1e4d2d3ac18745460e77052488b83a8dc80e7a/tools/mdm/migration/micromdm/touchless/main.go#L277

That being said... it may be somewhat useful as a debugging tool, to see that some host did phone in. Now that this work is done, maybe it's worth keeping it if only for that?

Reply:
I'm wondering if part of the reason why the batch operations crush the writer may be because we send a big batch of push notifications, which leads to a thundering herd of `StoreCommandReport` with `Idle` statuses, followed by another herd of `Acknowledged` statuses. I think this would be especially problematic in load test where osquery-perf means there's no latency of APNS (although not sure what APNS latency looks like IRL). 

Plus all those row locks probably spill over into the other operations like `RetrieveNextCommand` that also show up in our top SQL waits. 

It seems like we might want to take a closer look at this whole [flow](https://vscode.dev/github/fleetdm/fleet/blob/22361-mdm-min-os-ade-sso/server/mdm/nanomdm/storage/mysql/queue.go#L115-L116) as well as how the way we're using pushes might play into things.

Reply:
> That being said... it may be somewhat useful as a debugging tool, to see that some host did phone in. Now that this work is done, maybe it's worth keeping it if only for that?

I think it could definitely be useful in debugging, especially if we make it clear to callers in the future that it is async. I wonder if we might want to drop the column in the queue and use this PR to add nano seen times as its own separate table?

Reply:
You sent a vscode-thing link that wants me to login, but I get what you mean. But I'm not sure what this means for this PR - the fact remains that this update is hammering the writer (it's not a batch operation - the fix is to batch operations but the original issue is a single host's last_seen_at getting updated). Maybe in real life it's less obvious due to what you mention - the push notifications latency - but it's likely that at scale those notifications lead to some kind of thundering herd too? Unless Apple takes care of that for us and spreads out those notifications for the same account, but that seems unlikely?

Reply:
>  I wonder if we might want to drop the column in the queue and use this PR to add nano seen times as its own separate table?

That's a somewhat significantly more involved change, would probably mean redoing the load tests and changing the touchless migration which I have no idea how we can test. I'd be hesitant to do that.

Reply:
I guess another option is to put that PR on hold until we have feedback of that update being problematic under load, since we believe it may be a side-effect of our `osquery-perf` implementation and its thundering herd-style check-in.

Reply:
Ah, sorry, here's a GH [link](https://github.com/fleetdm/fleet/blob/7a595d51e471402778d909e53964369dc12f909c/server/mdm/nanomdm/storage/mysql/queue.go#L115-L116). 

> it's not a batch operation - the fix is to batch operations but the original issue is a single host's last_seen_at getting updated

That's right, I should have been more clear. I meant batch in the sense of `fleetctl gitops` doing changes to "batches" of profiles, which then leads to the single row updates of `last_seen_at`. We see the problem when only one profile changes on a team via the UI too. But it is especially bad in gitops scenarios.

> But I'm not sure what this means for this PR

It's definitely an improvement to the existing situation. I do think we can keep this PR as it is and look at the root problem separately. I am a little worried that we have a cluster of issues (this one, Victor's issues, and others) where we're adding complexity around certain queries in isolation but not taking a more wholistic view, which might make it more difficult to untangle in the future. I think we are in a good position right now to appreciate some fundamental bottlenecks in terms of how we're using nano. I'd hate to see us lose the thread and not follow up on the bigger picture.

Reply:
> I am a little worried that we have a cluster of issues (this one, Victor's issues, and others) where we're adding complexity around certain queries in isolation but not taking a more holistic view, which might make it more difficult to untangle in the future. I think we are in a good position right now to appreciate some fundamental bottlenecks in terms of how we're using nano. I'd hate to see us lose the thread and not follow up on the bigger picture.

That's a great point. How about this to unblock things:

* Merge this fix
* When ready merge Victor's fix
* Create a new eng-initiated issue for the more holistic view/improvement of the flow
* Let it wait a bit before tackling it (after the unified queue work might make sense), so we can a) see if there are still bottlenecks around this and b) see if the unified queue work impacts this flow in any way (I don't think it would, but it's such a big change around this that it might)

Reply:
Found this about APNS delays (iPhone thread, but probably similar for macOS): https://forums.developer.apple.com/forums/thread/105762

Reply:
Created the follow-up ticket: https://github.com/fleetdm/fleet/issues/23976

Reply:
Interesting, I wonder how APNS service levels compare for garden-variety apps vs. MDM services. Food for thought :)